community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java

```diff
@@ -0,0 +1,132 @@
+        {
+            int writtenOffset = 0;
+            while ( writtenOffset < trailer.length )
+            {
+                pageCursor.next();
+                pageCursor.setOffset( (int) ((writtenOffset + trailerOffset) % pagedFile.pageSize()) );
+                do
+                {
+                    do
+                    {
+                        pageCursor.putByte( trailer[writtenOffset] );
+                    }
+                    while ( pageCursor.shouldRetry() );
+                    writtenOffset++;
+                }
+                while ( (writtenOffset + trailerOffset) % pagedFile.pageSize() != 0 &&
+                        writtenOffset < trailer.length );
+            }
+        }
+    }
+
+    public static String readVersion( PagedFile pagedFile, String expectedVersion ) throws IOException
+    {
+        int encodedExpectedVersion = UTF8.encode( expectedVersion ).length;
+        String version = null;
+        int maximumNumberOfPagesVersionSpans = getMaximumNumberOfPagesVersionSpans( encodedExpectedVersion, pagedFile );
+        long pageId = Math.max( pagedFile.getLastPageId() + 1 - maximumNumberOfPagesVersionSpans, 0 );
+        try ( PageCursor pageCursor = pagedFile.io( pageId, PagedFile.PF_SHARED_LOCK ) )
+        {
+            byte[] allData = getLastBytes( pagedFile, maximumNumberOfPagesVersionSpans, pageCursor );
+            int offset = findTrailerOffset( allData, UTF8.encode( expectedVersion.split( " " )[0] ) );
+            if ( offset != -1 )
+            {
+                version = new String( allData, offset, encodedExpectedVersion, Charsets.UTF_8 );
+            }
+        }
+        return version;
+    }
+
+    public static int findTrailerOffset( byte[] src, byte[] sought )
+    {
+        for ( int i = src.length - sought.length; i >= 0; i-- )
+        {
+            int pos = 0;
+            while ( pos < sought.length && src[i + pos] == sought[pos] )
+            {
+                pos++;
+            }
+            if ( pos == sought.length )
+            {
+                return i;
+            }
+
+        }
+        return -1;
+    }
+
+    private static int getMaximumNumberOfPagesVersionSpans( int trailerLength, PagedFile pagedFile )
+    {
+        return trailerLength / pagedFile.pageSize() + 2;
+    }
+
+    private static byte[] getLastBytes( PagedFile pagedFile, int maximumNumberOfPagesVersionSpans,
+            PageCursor pageCursor ) throws IOException
+    {
+        byte[] allData = new byte[pagedFile.pageSize() * maximumNumberOfPagesVersionSpans];
+        int currentPage = 0;
+        while ( pageCursor.next() )
+        {
+            byte[] data = new byte[pagedFile.pageSize()];
+            do
+            {
+                pageCursor.getBytes( data );
+            }
+            while ( pageCursor.shouldRetry() );
+            System.arraycopy( data, 0, allData, currentPage * data.length, data.length );
```

Comment:
We can avoid this copy if we add a `getBytes( byte[] data, int writeOffset)` on the `PageCursor`